### ISSUE TITLE
Client Authentication Data Holder APIs called by the Register

### ISSUE THREAD

**Issue by commbankoss** [cdr-register/register#139]
**Background:**
The original ACCC design presented last year had the registry following an established pattern (same pattern is used for Data Recipients consuming Data Holders APIs):

1. Register application presents private_key_jwt to Data Holders’ Authorisation Server (AS)/IDP and receive access token;
2. Register application presents access token to Data Holder’s APIs (Resource Server).

At some point later, the design was pivoted to Register application presenting private_key_jwt / “signed jwt” directly to Data Holder’s APIs (Resource Server).
This shifted the responsibility for authentication from the AS / IDP to the API, which introduces additional complexity and security risks for Data Holder implementations.

**Change request:**
Consider this change in the next batch of Registry updates (release 1.3.0) and revert the pattern back to the original option which is in line with ‘private_key_jwt’ original intent:
1. Register application presents private_key_jwt to Data Holders’ Authorisation Server (IDP) and receive access token;
2. Register application presents access token to Data Holder’s APIs (Resource Server).

Preferably, this change should be done in time for the new Data Holders obligation date to simplify their implementation and improve ecosystem security posture (July 2021).

**Comment by CDR-Register-Stream** [comment]
@commbankoss Thank you for your query. I'll dig through the history of this issue so we can understand current state. 

As for the change request, DSB and ACCC will look at this issue to understand the complexity and security risks.

Please stay tuned

**Comment by CDR-Register-Stream** [comment]
We have identified options on how the ecosystem could manage this issue.

**Option 1:**
No change. Maintain the `signed_jwt` authentication method for current and future versions of GetMetrics API.

**Option 2:**
We have a cut-over of authentication where newer versions of the APIs support the presentation of access tokens
- V1 of the GetMetrics API would conform to the `signed_jwt` authentication method
- V2+ of the GetMetrics API would introduce issuance of access tokens at the IDP (to the CDR Register) and presentation of the access token to the GetMetrics API

**Option 3:**
All API versions are changed to support issuance of access tokens at the IDP (to the CDR Register) and presentation of the access token to the GetMetrics API. 

**Option 4:**
We rely on the discovery mechanism. The CDR-Register will look-up the DH [OpenID Configuration End Point](https://consumerdatastandardsaustralia.github.io/standards/#end-points), identify which authentication method is being exposed and uses this accordingly.

---

For Options 2-4, further work would be required to define the registration mechanism for the CDR Register within the Data Holder IDPs and the definition of authorisation scopes to call Admin APIs.

We are keen to provide certainty on this so data holders in the 2nd tranche of the CDR can reduce the scope of their build and current data holders can plan build/migration

**Comment by ryderwj** [comment]
The current model where the signed jwt is presented directly to the API required a custom approach for our implementation to manage the authentication, rather than delegating this to the IDP.  Build work has been completed for this. 

Options 2-4 allow for a more "standard" approach but will require a slightly different registration process for the CDR Registry.  No specific preferences within these options as the implications are more related to those already live with the metrics endpoint.  All would mean additional work is required.

**Comment by perlboy** [comment]
There seems little reason to not align authentication with the Admin API according to international standards related to using `client_credentials` with a `private_key_jwt` to a designated IDP with an access token then presented to the resource server. This would, in essence, avoid a specific build for the Admin APIs and follow norms for retrieving data. As a party involved in active implementations Biza.io would prefer this be resolved in line with the Get Metrics V2 call so that we aren't immediately faced with a deprecated API.

**Comment by CDR-Register-Stream** [comment]
@ryderwj Thanks for your input. 

We are looking at a model where we can provide flexibility to Data Holders so alignment to the international standards does not impose mandatory build requirements for current delivery time frames.

**Comment by CDR-Register-Stream** [comment]
All Data Holders which implement the ``client credentials`` with ``private_key_jwt`` authentication will require a registration approach to be defined.

We have some options on how the CDR Register may be registered with Data Holders to facilitate this authentication approach.

Options at hand are:
1. Treat the CDR Register as any other client, requiring them to go through the DCR process. The DCR process for the Register would need to be explicitly defined as the current definition is specified for Software Product registration
2. Registration is the responsibility of Data Holders who then update the CDR Register with an issued client ID. Data Holders are already aware of the CDR Register JWKS URI given the current authentication model. 
3. Require Data Holders to use a fixed client ID: **cdr-register** 

I would be keen to get feedback from the community on what Registration approach is feasible and then we can flesh out the details

**Comment by perlboy** [comment]
Firstly, from my perspective all of these options are workable because at least the ACCC would be accessing an API "normally" with a Bearer token. 

Working through each one though:
>Treat the CDR Register as any other client, requiring them to go through the DCR process. The DCR process for the Register would need to be explicitly defined as the current definition is specified for Software Product registration

This is a classic "dog fooding" solution and to me this seems like the most natural and would involve the introduction of a scope (for instance `cdr:administration`) which could be folded into the SSA using the existing processes. Registering the ACCC as a software product seems entirely legitimate considering the ACCC is consuming APIs.

>Registration is the responsibility of Data Holders who then update the CDR Register with an issued client ID. Data Holders are already aware of the CDR Register JWKS URI given the current authentication model.

This is workable and avoids static client id complexity below but will also mean that IDP changes impact Register and split brain can occur during this (ie. Register authenticating to a new IDP prior to cut over).

> Require Data Holders to use a fixed client ID: cdr-register

This would not be preferred as it means any testing infrastructure for Holders would need to have the same client id registered and multi-tenant/multi-brand solutions could not separate security concerns between client identifiers.

**Comment by bmangalaganesh** [comment]
Agree with the thoughts voiced by @perlboy above.

Prefer Option #1 - Using the DCR process for CDR Register. This approach keeps things simple (and elegant too!)

**Comment by NationalAustraliaBank** [comment]
NAB supports Option 1

> 1. Treat the CDR Register as any other client, requiring them to go through the DCR process. The DCR process for the Register would need to be explicitly defined as the current definition is specified for Software Product registration

**Comment by CDR-Register-Stream** [comment]
@perlboy, @bmangalaganesh, @NationalAustraliaBank thanks for your input.

With my review of **Option 1**, I am struggling to see an elegant solution and I'd be keen with collaborative input on this.

The DCR process currently in place is designed to allow DHs to trust ADR Software Products using the signed SSA to convey this trust. The metadata described in the [SSA definition](https://cdr-register.github.io/register/#dynamic-client-registration) and the [registration request using JWT](https://cdr-register.github.io/register/#registration-request-using-jwt) contains several mandatory fields which would not be relevant for the CDR Register registering with the DH. Examples include `redirect_uris` described in the SSA. This field would either need to be made optional or we would need to explicitly define the behaviour for passing an empty array.

If we extrapolate this out, we would either need to modify the [SSA ](https://cdr-register.github.io/register/#dynamic-client-registration)and [registration request using JWT](https://cdr-register.github.io/register/#registration-request-using-jwt) to handle both use cases (Software Products and the CDR Register) OR we duplicate the process and optimise it for the CDR Register.

I am keen on ensuing that we align to the DCR process to avoid build requirements for the DHs. However, redefining the registration request process and imposing this on the DH does not map to this philosophy.

The CDR Register is a client which will exist for the known future of the regime. **Option 2** allows for Data Holders to enter the CDR Register into their IDPs using their own processes. We don't need to duplicate the design or heavily augment the current one to cater for a use case which each DH will execute only once as part of on-boarding to the CDR.

I'd be keen to understand from Data Holders how onerous they see **option 2** is compared to **option 1** as I currently see significant overhead in **option 1**

**Comment by perlboy** [comment]
As discussed in the call, on review and in the interest of simplifying the approach and reaching a target state as soon as possible, we are comfortable with Option 1 or 2 but acknowledge that Option 2 is the simplest to implement and therefore would be happy to do it.

**Comment by CDR-Register-Stream** [comment]
Thanks everyone for their input.

Based on feedback received so far and weighing up where the complexity lies, ACCC's current position is as follows.

1. Data Holders can optionally implement the `client credentials` with `private_key_jwt` authentication mechanism
2. Data Holders which use this approach will be required to add the CDR Register to their IDPs using their own processes. Data Holders will then issue the CDR Register a `client id` and distribute this to the ACCC as part of onboarding
3. The CDR Register will retrieve GetMetrics using `signed-jwt` OR `client credentials` with `private_key_jwt` authentication, depending on whether a registration has been received by the CDR Register

We are keen to bring this issue to completion so invite further feedback on this position until 15th November 2020 where we will then make a final decision.

**Comment by commbankoss** [comment]
Commonwealth Bank supports aligning this interaction with the normative standards.
We would like to make two suggestions:
1. The target state (client_credentials with private_key_jwt) should be outlined, including the transition period where the old and new methods co-exist as to allow transition time for participants that have already implemented an old method.
2. Regarding provisioning of client ID, we prefer dynamic client registration for greater consistency across ecosystem. We believe that, for some participants, the effort required to manually provision the client would be similar to the effort to update current dynamic client registration. This will also also future proof the setup in the case future required data / APIs / scopes would require other scopes, and other entities on behalf of registry might require access to some reporting (e.g. Treasury).

**Comment by NationalAustraliaBank** [comment]
NAB does not support having a transient state (where there is an option for DHs to choose the authentication mechanism on the GET Metrics) without a transition plan, as it is already acknowledged that the target state is to move to an authentication mechanism which uses client_credentials and private_key_jwt at the token endpoint to get access token and presenting the access token to the GET Metrics API. 
 
With the registration of the CDR Register, we prefer Option 1 i.e., using Dynamic client registration as a target state. In the transition state, we support the option to create the client_id using our internal processes, but the instructions to provide the client id to CDR Register should be defined along with the discovery of jwks endpoint etc.  
 
We expect a transition plan to the target state to be outlined as part of this change.

**Comment by CDR-Register-Stream** [comment]
ACCC's current position is that **option 2** for client registration is the only model which is feasible for this functionality to be available in the July 2021 timeframe.  Therefore, to be clear, the register design will document the following:

_Data Holders which use this approach will be required to add the CDR Register to their IDPs using their own processes. Data Holders will then issue the CDR Register a client id and distribute this to the ACCC as part of onboarding_

We have raised issue #[360](https://github.com/ConsumerDataStandardsAustralia/standards-maintenance/issues/360) on the CDS standards maintenance repo to take this issue through the change management process. I'd encourage participants to document their opinions on this newly raised issue so that DSB can quantify support.

**Comment by CDR-Register-Stream** [comment]
Issue #[360](https://github.com/ConsumerDataStandardsAustralia/standards-maintenance/issues/360) addresses authentication mechanisms for data holder admin APIs and the registration processes. 

The CDR Register design defers to the [Client Authentication](https://consumerdatastandardsaustralia.github.io/standards/#client-authentication) section in the Consumer Data Standards to articulate these mechanisms. This work will be completed as part of CDS Maintenance Iteration 6.

The CDR Register will support both the `signed-jwt` and `private_key_jwt` authentication mechanisms prior to the July 31st GetMetrics obligations for 2nd tranche data holders. Data holders can choose which method is appropriate for delivery in this timeframe.

Client_ids generated for the CDR Register will be captured and stored on the Register via the onboarding process.

**Comment by CDR-Register-Stream** [comment]
[V1.4.0](https://cdr-register.github.io/register/#client-authentication) of the CDR Register design defers client authentication specification to the Consumer Data Standards. 

The above proposed change is being managed through [issue 360](https://github.com/ConsumerDataStandardsAustralia/standards-maintenance/issues/360)
